File: C 26.08.2022/TestSuite.c
/*
  The random test suite is developed by the following CodeWars user: www.codewars.com/users/porkfreezer
*/

#include <criterion/criterion.h>
#include <unistd.h>
#include <time.h>
#include <stdio.h>

char *find_the_number_plate(int id);

static void do_test(int id, const char *plate) {
  char *actual = find_the_number_plate(id);
  cr_assert_str_eq(plate, actual, "Expected %s for %d, but got %s", plate, id, actual);
  if (actual)
    free(actual);
}

Test(number_plates, should_pass_random_tests) {
  srand(time(NULL));
  for (int i = 0; i < 100; i++) {
    int input = rand() % (999*26*26*26);

    char expected[7];
    int id = input;
    int sn = id % 999 + 1;
    char l1 = (id /= 999) % 26 + 'a';
    char l2 = (id /= 26) % 26 + 'a';
    char l3 = (id / 26) % 26 + 'a';
    sprintf(expected, "%c%c%c%03d", l1, l2, l3, sn);

    do_test(input, expected);
  }
}

Test(number_plates, should_pass_basic_tests) {
  do_test(3, "aaa004");
  do_test(1487, "baa489");
  do_test(40000, "oba041");
  do_test(17558423, "zzz999");
  do_test(10598400, "bsp010");
  do_test(234567, "aja802");
  do_test(43056, "rba100");
}

File: Go 19.08.2022/TestSuite.go
/*
  The random test suite is developed by the following CodeWars user: www.codewars.com/users/akar-0
*/

package kata_test
import (
  . "github.com/onsi/ginkgo"
  . "github.com/onsi/gomega"
  . "codewarrior/kata"
  "math/rand"
  "time"
  "math"
  "fmt"
)

func referenceSolution(n int) string {
  var res []rune
  for _,i := range []float64{2.0,1.0,0.0} {
    d := 999 * int(math.Pow(26.0, i))
    l := n / d
    n %= d
    res = append([]rune{rune(l+97)}, res...)
  }
  return string(res) + fmt.Sprintf("%03d", n+1)
}

func dotest(n int, expected string) {
     Expect(FindTheNumberPlate(n)).To(Equal(expected), "With n = %d", n)
}

var _ = Describe("Tests", func() {
     It("Sample tests", func() {
        dotest(3,"aaa004")
        dotest(1487,"baa489")
        dotest(40000,"oba041")
        dotest(17558423,"zzz999")
        dotest(43056,"rba100")
        dotest(234567,"aja802")
        dotest(10598400,"bsp010")
     })
     It("Random tests", func() {
       rand.Seed(time.Now().UTC().UnixNano())
       for i := 0 ; i < 100 ; i++ {
         n := rand.Intn(17558424)
         dotest(n, referenceSolution(n))
       }
     })
})
